### ISSUE TITLE
Can't build locally — no package-lock.json

### ISSUE THREAD

**Issue by matthias-ccri** [openlayers/ol-cesium#962]
Hello,

I attempted to pull the repo today and run `make dist`. I got an error:

```
$ make dist
npm run doc

> ol-cesium@2.13.0 doc /opt/devel/src/ol-cesium
> typedoc --name ol-cesium --excludeExternals --out dist/apidoc --entryPoints src/index.library.js --tsconfig ./tsconfig.json

Warning: You are running with an unsupported TypeScript version! TypeDoc supports 3.9, 4.0, 4.1, 4.2
TypeDoc exiting with unexpected error:
TypeError: Cannot read property 'kind' of undefined
    at Object.isDeclaration (/opt/devel/src/ol-cesium/node_modules/typescript/lib/typescript.js:13449:18)
    at Object.isDeclarationName (/opt/devel/src/ol-cesium/node_modules/typescript/lib/typescript.js:16414:75)
    at getTypeOfSymbolAtLocation (/opt/devel/src/ol-cesium/node_modules/typescript/lib/typescript.js:66938:20)
    at Object.getTypeOfSymbolAtLocation (/opt/devel/src/ol-cesium/node_modules/typescript/lib/typescript.js:45176:35)
    at Object.convertProperty (/opt/devel/src/ol-cesium/node_modules/typedoc/dist/lib/converter/symbols.js:330:176)
    at convertSymbol (/opt/devel/src/ol-cesium/node_modules/typedoc/dist/lib/converter/symbols.js:80:79)
    at convertSymbols (/opt/devel/src/ol-cesium/node_modules/typedoc/dist/lib/converter/symbols.js:86:9)
    at Object.convertClassOrInterface (/opt/devel/src/ol-cesium/node_modules/typedoc/dist/lib/converter/symbols.js:264:5)
    at convertSymbol (/opt/devel/src/ol-cesium/node_modules/typedoc/dist/lib/converter/symbols.js:80:79)
    at Object.convertAlias (/opt/devel/src/ol-cesium/node_modules/typedoc/dist/lib/converter/symbols.js:380:9)
npm ERR! code ELIFECYCLE
npm ERR! errno 5
npm ERR! ol-cesium@2.13.0 doc: `typedoc --name ol-cesium --excludeExternals --out dist/apidoc --entryPoints src/index.library.js --tsconfig ./tsconfig.json`
npm ERR! Exit status 5
npm ERR! 
npm ERR! Failed at the ol-cesium@2.13.0 doc script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/user/.npm/_logs/2021-06-16T18_08_51_702Z-debug.log
Makefile:39: recipe for target '.build/jsdoc.timestamp' failed
make: *** [.build/jsdoc.timestamp] Error 5
```

Please commit a working package-lock.json to the repo if you have one. This error likely arose because of versions drifting over time. Lockfiles should **always** be committed, because it prevents build failures like this.

Some people think library repos shouldn't commit lockfiles because libraries should always use the latest versions of everything. However, consider that all contributors _still have lockfiles_ in their local working copies. Do they all regenerate their lockfiles regularly? Probably not. So everyone has different lockfiles. The way to make a library stay current is to use CI jobs like greenkeeper or some similar way of regenerating the lockfile regularly. Just gitignoring the lockfile doesn't achieve the goal, because everyone *still has a lockfile*, but it's worse because they're all different lockfiles. With different lockfiles, the build may work on one machine but not on another (like mine).

After committing a lockfile to the repo, you can still regenerate it as usual (`rm -rf package-lock.json node_modules && npm i`). But this way the regenerations are tracked and everyone has the same lockfile.

Releasing seems to be a manual process according to RELEASE.md. Do you have plans to plug in some CI jobs? I can help with that, though I'm used to GitLab CI instead of GitHub Actions.

**Comment by gberaudo** [comment]
Hello @matthias-ccri, thank you for your message and your proposition for help.

The goal of this repo is to work with as many versions of OpenLayers and CesiumJS as possible, be it the latest version, or not.
Having the latest version of the dev dependencies is not a necessity either.

In this repository, we are basically following the CesiumJS policy to ignore lockfiles as their content depend on the node version, tend to create merge conflicts and it usually works great without them.
I have created a PR to remove them completly (same as what is done with CesiumJS).

A manual check is needed before releasing, a CI would not help there, but thanks for proposing.

**Comment by matthias-ccri** [comment]
@gberaudo I didn't respond earlier, but I should bring this up again.

Please reconsider committing the lockfile to this repo. If I may speak strongly and directly: Committing the lockfile is a best practice agreed upon by everyone including npm, yarn, and everyone who understands the purpose of a lockfile. If CesiumJS doesn't commit their lockfile then they are absolutely wrong and putting their code at risk with and gaining no benefit.

If since Jun 2021 you have since revised your opinion on lockfiles then good! But if not, I'd like to open this discussion with you. 

Some resources for reading:
- Npm docs saying in no uncertain terms to commit the lockfile: https://docs.npmjs.com/cli/v6/configuring-npm/package-locks
  > t is highly recommended you commit the generated package lock to source control: this will allow anyone else on your team, your deployments, your CI/continuous integration, and anyone else who runs npm install in your package source to get the exact same dependency tree that you were developing on. Additionally, the diffs from these changes are human-readable and will inform you of any changes npm has made to your node_modules, so you can notice if any transitive dependencies were updated, hoisted, etc.
  Note that this is the exact issue that caused me to create this GitHub Issue: I couldn't build your project because there was no lockfile. 
- A StackOverflow post with lots of great answers: https://stackoverflow.com/questions/44206782/do-i-commit-the-package-lock-json-file-created-by-npm-5
- Yarn likewise says clearly: "Lockfiles should be committed on all projects"
: https://classic.yarnpkg.com/blog/2016/11/24/lockfiles-for-all/

To respond to your points individually:
> The goal of this repo is to work with as many versions of OpenLayers and CesiumJS as possible, be it the latest version, or not.
Having a lockfile in git doesn't change this. You know, your computer already has a lockfile because npm creates it. So you are not avoiding having a lockfile, you are just ensuring that every developer has a different lockfile and you can't track changes between lockfiles. That's the most chaotic and disorganized option. If you really want to track changes to dependencies, then commit the lockfile, otherwise there's no way to track failing or passing dep combinations in git.

It's fine for a lockfile to change. You probably change the lockfile when you test older versions of deps. Just commit the changes, at least then they'll be tracked in git. It's no harm to commit lockfile changes.

> In this repository, we are basically following the CesiumJS policy to ignore lockfiles
I can't find official guidelines on this. Is this still official Cesium policy?

> lockfile contents depend on the node version
Not the version of `node` but the version of `npm` would be what causes discrepancies. However I haven't confirmed this claim. Do you have a reference for which versions of npm are incompatible with recent versions? 

But even if so, it's a tradeoff worth making. The workaround is to require contributors to use a compatible version of npm either upgrade or use a tool like [nvm](https://github.com/nvm-sh/nvm) to switch per project.

> tend to create merge conflicts
Merge conflicts are not an issue in lockfiles, if people know how to handle them (and they probably already do, because committing lockfiles is a standard practice). Lockfile conflicts are solved simply by running `npm` or `yarn` again. Yarn automatically resolves conflicts, and npm does too (see https://docs.npmjs.com/cli/v6/configuring-npm/package-locks#resolving-lockfile-conflicts).

> it usually works great without them
That word "usually" could be "always" if the lockfile is committed to git. This very ticket is proof of a case where _it doesn't work great without them_. None of the stated downsides outweighs the _instability_ of not having a lockfile. 

In short, here is why you should always commit lockfiles:
- lockfiles provide consistency of a single commit across time. Without it, packages in the npm registry will change, and your commit's _result_ will change with them, potentially breaking.
- lockfiles are snapshots of your node_modules. By analogy: a **git hash** is to **files** what a **lockfile** is to **node_modules**. You use git hashes to guarantee stability of file contents, and you use lockfiles to guarantee stability of node_modules (for a given commit). If you say "but node_modules should change so I can build against different deps" then you've missed the point. node_modules can change, just like files in git can change. The point is that you commit the hash after changing the files.

**Comment by matthias-ccri** [comment]
See also: https://stackoverflow.com/questions/44206782/do-i-commit-the-package-lock-json-file-created-by-npm-5/76058921#76058921

**Comment by matthias-ccri** [comment]
Resolved with https://github.com/openlayers/ol-cesium/pull/1040, thanks
